### DISCUSSION TITLE
How to limit / filter kafka metrics exposed over Prometheus endpoint

### DISCUSSION THREAD

**Discussion by sgv0007** [strimzi/strimzi-kafka-operator#10372] [Q&A] [answered]
Hi All ,

We need to filter or limit kafka metrics which we don't  monitor but unable to find where to add the filter criteria . We have a third part app scraping metrics from prometheus endpoint and they want filter logic to be applied at kafka broker level as they don't want to do that after reading all the metrics .

Thanks,
Shiva

**Comment by scholzj** [comment]
it depends a bit what metrics are you talking about. You can configure the metrics exposed through the JMX Prometheus Exporter in the Config Map that configures it. For Kafka Exporter, you can configure some things such as topic or group filters in the Kafka CR.

**Reply by sgv0007** [reply]
Thanks @scholzj  

If my understanding is correct then  in kafka metrics config map , If I provide exact metric name instead of pattern it would emit only that particular metric right?

rules:
# Special cases and very specific rules
- pattern : kafka_log_log_logstartoffset
  name: kafka_log_log_logstartoffset

**Reply by scholzj** [reply, marked as answer]
I'm not sure it is that simple as you are matching the JMX metrics and translating them into Prometheus metrics. But essentially yes, the rules there control what is exposed. So by replacing some of the wildcards with more specific values, you reduce the exposed metrics.

**Reply by sgv0007** [reply]
@scholzj  I was able to filter the metrics for kafka and kafka connect using config map but not able to do that same for Kafka Exporter . Any other option apart from filtering topicnames and groups?

**Reply by scholzj** [reply]
I'm not aware of any other way to filter something. As far as I know you can drop some metrics on the Prometheus side as well when scraping them.

**Reply by Yedidyah27** [reply]
@sgv0007 can you share how you dropped metrics in the configmap. I am trying to do the same without any success.
Added the following to the rules but seems not to be working.
```
- pattern: 'kafka_consumer_node_(incoming_byte|outgoing_byte|request|request_latency_avg|request_size_avg|response)'
      action: drop
    - pattern: 'kafka_consumer_io_ratio'
      action: drop
    - pattern: 'kafka_connect_node_(incoming_byte|outgoing_byte|request|request_latency_avg|request_size_avg|response)'
      action: drop
    - pattern: 'kafka_connect_io_(ratio|time_ns|time_ns_avg|wait_ratio|wait_time_ns|wait_time_ns_avg|waittime)'
      action: drop
    - pattern: 'kafka_producer_buffer(pool_wait_time_ns|pool_wait_time|pool_wait_ratio|_exhausted|_available_bytes|)'
      action: drop
    - pattern: 'kafka_admin_client_(successful_authentication|successful_authentication_no_reauth|successful_reauthentication|)'
      action: drop
```
